engine/datahistory_manager.go

```diff
@@ -0,0 +1,870 @@
+	if err != nil {
+		return err
+	}
+
+	if job.ID == uuid.Nil {
+		job.ID, err = uuid.NewV4()
+		if err != nil {
+			return err
+		}
+	}
+	if !toUpdate {
+		m.jobs = append(m.jobs, job)
+	}
+
+	dbJob := m.convertJobToDBModel(job)
+	return m.jobDB.Upsert(dbJob)
+}
+
+func (m *DataHistoryManager) validateJob(job *DataHistoryJob) error {
+	if job == nil {
+		return errNilJob
+	}
+	if !job.Asset.IsValid() {
+		return asset.ErrNotSupported
+	}
+	if job.Pair.IsEmpty() {
+		return errCurrencyPairUnset
+	}
+	exch := m.exchangeManager.GetExchangeByName(job.Exchange)
+	if exch == nil {
+		return fmt.Errorf("%s %w", job.Exchange, errExchangeNotLoaded)
+	}
+	pairs, err := exch.GetAvailablePairs(job.Asset)
+	if err != nil {
+		return err
+	}
+	if !pairs.Contains(job.Pair, false) {
+		return errCurrencyPairInvalid
+	}
+	if job.StartDate.After(job.EndDate) ||
+		job.StartDate.IsZero() ||
+		job.EndDate.IsZero() ||
+		job.StartDate.After(time.Now()) {
+		return errInvalidTimes
+	}
+	if job.Results == nil {
+		job.Results = make(map[time.Time][]DataHistoryJobResult)
+	}
+
+	if job.RunBatchLimit <= 0 {
+		log.Warnf(log.DataHistory, "job %s has unset batch limit, defaulting to %v", job.Nickname, defaultBatchLimit)
+		job.RunBatchLimit = defaultBatchLimit
+	}
+	if job.MaxRetryAttempts <= 0 {
+		log.Warnf(log.DataHistory, "job %s has unset max retry limit, defaulting to %v", job.Nickname, defaultRetryAttempts)
+		job.MaxRetryAttempts = defaultRetryAttempts
+	}
+
+	if job.DataType == dataHistoryTradeDataType &&
+		job.Interval <= 0 {
+		job.Interval = defaultTradeInterval
+	}
+
+	b := exch.GetBase()
+	if !b.Features.Enabled.Kline.Intervals[job.Interval.Word()] {
+		return fmt.Errorf("%s %w unsupported by exchange %s", job.Interval.Word(), kline.ErrUnsupportedInterval, job.Exchange)
+	}
+
+	job.StartDate = job.StartDate.Round(job.Interval.Duration())
+	job.EndDate = job.EndDate.Round(job.Interval.Duration())
+
+	return nil
+}
+
+// GetByID returns a job's details from its ID,
+func (m *DataHistoryManager) GetByID(id string) (*DataHistoryJob, error) {
+	if m == nil {
+		return nil, ErrNilSubsystem
+	}
+	if atomic.LoadInt32(&m.started) == 0 {
+		return nil, ErrSubSystemNotStarted
+	}
+	m.m.Lock()
+	for i := range m.jobs {
+		if m.jobs[i].ID.String() == id {
+			cpy := m.jobs[i]
+			m.m.Unlock()
+			return cpy, nil
+		}
+	}
+	m.m.Unlock()
+	dbJ, err := m.jobDB.GetByID(id)
+	if err != nil {
+		return nil, fmt.Errorf("%w with id %s %s", errJobNotFound, id, err)
+	}
+	result, err := m.convertDBModelToJob(dbJ)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert model with id %s %w", id, err)
+	}
+	return result, nil
+}
+
+// GetByNickname searches for jobs by name and returns it if found
+// returns nil if not
+// if fullDetails is enabled, it will retrieve all job history results from the database
+func (m *DataHistoryManager) GetByNickname(nickname string, fullDetails bool) (*DataHistoryJob, error) {
```

Comment:
![image](https://user-images.githubusercontent.com/31364354/120760605-a2421b00-c557-11eb-8e5d-51a03e2059ec.png)
I am unsure where to post this nit but maybe that should come up with an ID not valid error or something like that

Reply:
I'll make it default to ID then if someone chooses not to add anything